src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs

```diff
@@ -2,25 +2,106 @@
+            NotModified,
+            ShouldProcess,
+            PreconditionFailed,
+        }
+
         protected ILogger Logger { get; }
 
-        protected void SetHeadersAndLog(ActionContext context, FileResult result)
+        protected (RangeItemHeaderValue range, long rangeLength, bool serveBody) SetHeadersAndLog(ActionContext context, FileResult result, long? fileLength, DateTimeOffset? lastModified = null, EntityTagHeaderValue etag = null)
         {
             SetContentType(context, result);
             SetContentDispositionHeader(context, result);
             Logger.FileResultExecuting(result.FileDownloadName);
+            if (fileLength.HasValue)
+            {
+                SetAcceptRangeHeader(context);
+            }
+
+            var request = context.HttpContext.Request;
+            var httpRequestHeaders = request.GetTypedHeaders();
+            var response = context.HttpContext.Response;
+            var httpResponseHeaders = response.GetTypedHeaders();
+            if (lastModified.HasValue)
+            {
+                httpResponseHeaders.LastModified = lastModified;
+            }
+            if (etag != null)
+            {
+                httpResponseHeaders.ETag = etag;
+            }
+
+            var preconditionState = CheckPreconditionHeaders(
+                    context,
+                    httpRequestHeaders,
+                    lastModified,
+                    etag);
+            var serveBody = true;
+            if (HttpMethods.IsHead(request.Method))
+            {
+                serveBody = false;
+            }
+            else
+            {
+                if (request.Headers.ContainsKey(HeaderNames.Range))
+                {
+                    if (preconditionState.Equals(PreconditionState.Unspecified) ||
+                        preconditionState.Equals(PreconditionState.ShouldProcess))
+                    {
+                        var rangeInfo = SetRangeHeaders(context, httpRequestHeaders, fileLength, lastModified, etag);
+                        if (response.StatusCode == StatusCodes.Status416RangeNotSatisfiable)
+                        {
+                            serveBody = false;
+                        }
+                        return (rangeInfo.range, rangeInfo.rangeLength, serveBody);
+                    }
+                }
+
+                if (preconditionState.Equals(PreconditionState.NotModified))
+                {
+                    serveBody = false;
+                    response.StatusCode = StatusCodes.Status304NotModified;
+                }
+                else if (preconditionState.Equals(PreconditionState.PreconditionFailed))
+                {
+                    serveBody = false;
+                    response.StatusCode = StatusCodes.Status412PreconditionFailed;
+                }
+            }
+
+            return (null, 0, serveBody);
+        }
+
+        private void SetContentType(ActionContext context, FileResult result)
```

Comment:
Not part of my change https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs#L41. Should I still make address the feedback?